### PR TITLE
Generate binary packages on Travis automatically v2

### LINKED ISSUE #711: Proposition: DRAKVUF binary builds
Binary builds for supported distributions would make it much easier to set everything up.

PoC build scripts: https://github.com/icedevml/drakvuf-deb/tree/master
PoC release of 0.6 for Ubuntu https://github.com/icedevml/drakvuf-deb/releases/tag/0.6

The build scripts are based upon Xen's `mkdeb` tool. I've decided to bundle everything together (i.e. Xen + LibVMI + DRAKVUF in a single package), as DRAKVUF is currently relying on particular versions of dependencies and it is very hard to support many different versions at once.

For now I have stable Ubuntu 18.04 build as well as Debian Buster one.

I think for now it is saving about 70% of time, mostly because there is no need to build Xen. I wonder how stable it is. Gonna probably use this way of distribution on our production.

### PR DESCRIPTION
**Pull request by icedevml**
Fixes #711

### CODE REVIEW

**Comment by icedevml** [comment]
Yet to be done:
* change prefix to `/usr`
* implement `apt-get build-dep`

**Comment by icedevml** [comment]
What I've implemented is a split to two Dockerfiles, one for Xen and one for the rest of the stuff. The intermediate image is saved with `docker save` and then cached using Travis CI. This should be pretty robust, as `docker save` would essentially save the full filesystem at the point when Xen was succesfully installed.

Work in progress.

File: package/build.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+if [ ! -f "package/cache/xen-intermediate-$XEN_HASH.tar.gz" ]
+then
+    docker build -f package/Dockerfile-xen -t xen-intermediate .
+    echo save
+    docker save xen-intermediate | gzip -c > "package/cache/xen-intermediate-$XEN_HASH.tar.gz"
+    echo done
+else
+    docker load < "package/cache/xen-intermediate-$XEN_HASH.tar.gz"
+fi
+echo omg
```
Comment (tklengyel) [inline package/build.sh]: :)

Comment (icedevml) [reply package/build.sh]: Yep, already fixed :P

**Comment by icedevml** [comment]
TODO error code checks

**Comment by tklengyel** [comment]
Also don't forget to add the Travis caching for `package/cache`.

**Comment by icedevml** [comment]
@tklengyel I wonder if these old `xen-intermediate*` would be deleted after some time or we need explicit cleanup?

**Comment by icedevml** [comment]
Ah got it

> There is one cache per branch

so one `xen-intermediate-*` is enough for any case

**Comment by tklengyel** [comment]
Right, + the `before_deploy` section is only triggered for builds on the `master` branch, ie. after a merge

File: package/Dockerfile-xen [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+ENV DEBIAN_FRONTEND=noninteractive
+ENV USER root
+
+# build depends
+RUN mkdir -p /log && \
+    sed -i 's/# deb-src/deb-src/g' /etc/apt/sources.list && \
+    apt-get update && \
+    echo Running apt-get install, this may take a few minutes... && \
+    apt-get --quiet --yes install wget git bcc bin86 gawk bridge-utils iproute2 libcurl4-openssl-dev bzip2 pciutils-dev build-essential make gcc clang libc6-dev libc6-dev-i386 linux-libc-dev zlib1g-dev libncurses5-dev patch libvncserver-dev libssl-dev iasl libbz2-dev e2fslibs-dev git-core uuid-dev ocaml libx11-dev bison flex ocaml-findlib xz-utils gettext libyajl-dev libpixman-1-dev libaio-dev libfdt-dev cabextract libglib2.0-dev autoconf automake libtool libjson-c-dev libfuse-dev liblzma-dev autoconf-archive kpartx python-dev cmake libsdl-dev && \
```
Comment (icedevml) [inline package/Dockerfile-xen]: For some reason, build fails when this long line gets replaced with `apt-get --quiet --yes install wget && apt-get --quiet --yes build-dep xen xen-tools`. TODO investigate further

Comment (tklengyel) [reply package/Dockerfile-xen]: The package is called `xen-hypervisor-4.9-amd64` on 18.04: https://packages.ubuntu.com/search?suite=bionic&searchon=names&keywords=xen-hypervisor

Comment (tklengyel) [reply package/Dockerfile-xen]: I guess you should do `build-dep xen-utils-4.9 xen-hypervisor-4.9-amd64`

Comment (icedevml) [reply package/Dockerfile-xen]: When I type `xen-utils-4.9 xen-hypervisor-4.9-amd64` it's telling something like "using `xen` as source package instead" for both of these, so it doesn't change anything at all.

Turns out that everything works if I additionally install `build-essential git wget flex bison` (+ LibVMI/DRAKVUF dependencies).

**Comment by icedevml** [comment]
Ok, ready to review. I think that all sub-issues were fixed.

File: .travis.yml [hunk 3]
```diff
@@ -257,3 +259,31 @@ matrix:
...
+        - TEST="package build"
+      script:
+        - echo "Building binary packages"
+      before_deploy:
+        - git submodule update --init libvmi xen drakvuf-doppelganging
+        - travis_wait 60 sh package/build.sh;
+        - git config --local user.name "drakvuf-builds"
+        - git config --local user.email "drakvuf-builds@tklsoftware.com"
+        - export TRAVIS_TAG=${TRAVIS_TAG:-$(date +'%Y%m%d%H%M%S')-$(git log --format=%h -1)}
+        - git tag $TRAVIS_TAG
```
[the remark this answers is not in the document]
Comment (icedevml) [reply .travis.yml]: Sure, I will add this

[the remark this answers is not in the document]
Comment (icedevml) [reply .travis.yml]: done, output from `make dist` will be also outputted to the `package/out/` directory, using exactly the same sources from which `.deb` was built

**Comment by tklengyel** [comment]
OK, I think we are good with merging this. We can tweak it later if needed.

**Pull request merged**